### PR TITLE
[Wallet] Do not perform ECDSA signing in the fee calculation inner loop.

### PR DESCRIPTION
**Pull request by gmaxwell**
Performing signing in the inner loop has terrible performance
 when many passes through are needed to complete the selection.

Signing before the algorithm is complete also gets in the way
 of correctly setting the fee (e.g. preventing over-payment when
 the fee required goes down on the final selection.)

Use of the dummy might overpay on the signatures by a couple bytes
 in uncommon cases where the signatures' DER encoding is smaller
 than the dummy: Who cares? (Probability ~= 1/256^(bytes/2.))

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK - I thought this was already the case!

**Comment by jonasschnelli** [comment]
Concept ACK.

**Comment by morcos** [comment]
Concept ACK, will review.  I was wondering about this myself.  This will also make #9404 even a smaller change as we won't need to track variance.

**Comment by sipa** [comment]
Concept ACK. I knew it wasn't the case, but it seems to obvious...

**Comment by morcos** [comment]
utACK aaa3cfc

**Review by dcousens — APPROVED** [review]

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2558,6 +2542,37 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
                 continue;
             }
         }
+
+        if (sign)
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: A nice by-product of this change is that the signing is factored out. Which means that at some point this huge function could be split into multiple clearly defined functions.

**Comment by laanwj** [comment]
utACK aaa3cfc

**Review by laanwj — APPROVED** [review]

**Comment by morcos** [comment]
@gmaxwell oh hmm...  You are now signing outside of cs_main, cs_wallet.  
I suppose that is ok?

**Comment by sipa** [comment]
I think keystore is thread safe (and that that's all that is needed for
signing).

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2500,26 +2495,15 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
-                wtxNew.SetTx(MakeTransactionRef(std::move(txNew)));
+                CTransaction txNewConst(txNew);
+                dPriority = txNewConst.ComputePriority(dPriority, nBytes);
 
-                // Limit size
-                if (GetTransactionWeight(wtxNew) >= MAX_STANDARD_TX_WEIGHT)
-                {
-                    strFailReason = _("Transaction too large");
-                    return false;
+                // Remove scriptSigs to eliminate dummy signatures for fee calculation
```
Comment (sipa) [inline src/wallet/wallet.cpp]: This is a bit confusing to me. We're not elimining the dummy signatures in order to do fee calculation, but removing the dummy signatures which were added in order to do fee calculation.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2558,6 +2542,37 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
                 continue;
             }
         }
+
+        if (sign)
+        {
+            CTransaction txNewConst(txNew);
+            int nIn = 0;
+            BOOST_FOREACH(const PAIRTYPE(const CWalletTx*,unsigned int)& coin, setCoins)
```
Comment (sipa) [inline src/wallet/wallet.cpp]: Feel like switching this to `for (const auto& coin : setCoins)` ?

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (sipa) [inline src/wallet/wallet.cpp]: Agree!

**Review by sipa — COMMENTED** [review]
utACK aaa3cfc485bc570a3e537349130a3471a8a4d3a8

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (gmaxwell) [inline src/wallet/wallet.cpp]: There were only a couple extra changes to do all of them and avoid making the function inconsistent in its usage, so I did. Fixed.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (gmaxwell) [inline src/wallet/wallet.cpp]: I reordered the sentence to make it more clear that I mean the "for fee estimation dummies". Should be fixed.

**Comment by gmaxwell** [comment]
@morcos 

> oh hmm... You are now signing outside of cs_main, cs_wallet. I suppose that is ok?

It's not outside of the lock unless I'm missing something!

**Comment by morcos** [comment]
@gmaxwell oops, yeah i was misreading braces..   

well maybe you should move it out?
:)

**Review by sdaftuar — APPROVED** [review]
utACK b3d7b1cbe7afdc6a63bbcbe938e8639deedb04a1

**Comment by sipa** [comment]
utACK b3d7b1cbe7afdc6a63bbcbe938e8639deedb04a1

**Pull request merged**